opm/autodiff/BackupRestore.hpp

```diff
@@ -0,0 +1,336 @@
+/*
+  Copyright 2014 IRIS AS
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef OPM_BACKUPRESTORE_HEADER_INCLUDED
+#define OPM_BACKUPRESTORE_HEADER_INCLUDED
+
+#include <iostream>
+
+#include <opm/core/simulator/SimulatorState.hpp>
+#include <opm/core/simulator/BlackoilState.hpp>
+#include <opm/autodiff/WellStateFullyImplicitBlackoil.hpp>
+
+namespace Opm {
+
+    namespace {
+        template <class T>
+        void writeValue( std::ostream& stream, const T& value )
+        {
+            stream.write( (const char *) &value, sizeof( T ) );
+        }
+
+        template <class T>
+        void readValue( std::istream& stream, T& value )
+        {
+            stream.read( (char *) &value, sizeof( T ) );
+        }
+
+        template <class Vector>
+        void writeVector( std::ostream& stream, const Vector& vector)
+        {
+            typedef typename Vector :: value_type T;
+            unsigned int size = vector.size() * sizeof( T );
+            writeValue( stream, size );
+            if( size > 0 ) {
+                stream.write( (const char *) vector.data(), size );
+            }
+        }
+
+        template <class Map>
+        void writeMap( std::ostream& stream, const Map& m)
```

Comment:
This function seems pretty specialized, requiring keys and values to be `Vector`s. I now learned that std::string has data() and size() which makes this work, that is a little bit obscured here, so a short comment would be nice to inform the next reader.